### DISCUSSION TITLE
Classification images b&w vs color

### DISCUSSION THREAD

**Discussion by Jonseed** [d8ahazard/sd_dreambooth_extension#129] [Q&A] [answered]
My class prompt produces black and white images by default, but the dataset I'm training is in color. Should I adjust my class prompt so it generates color images?

**Comment by d8ahazard** [comment, marked as answer]
I don't know if there's a definitive answer to this, but this is my thought process:

The classification images are sort of a baseline or control for what the new subject images should look like in the most generic sense possible. Specifically, what the existing model was creating before training started, and what it's creating now.

I've read a few guides that suggest just taking random datasets from the web and using those as classifiers, but from my most basic and unscientific tests, I haven't gotten great results this way.

In contrast, just generating the basic class images using the existing model seems to provide better editability and coherence.

Again, none of this is science, just me tinkering with one dataset, but from what I can tell, just let the prompt do whatever it's going to do.

Alternatively, I'd be curious to know what would happen if we generated a wider variety of class images with custom prompts embedded in the filename/txt, and using those prompts with the class image. I'll probably implement that at some point in time in the future....

**Reply by YukihimeX** [reply]
Training images with caption (aka image & text generated/processed by BLIP) is the same approach as the [Pokemon Stable Diffusion](https://github.com/LambdaLabsML/examples/tree/main/stable-diffusion-finetuning) I read about but it would be great if you can implement it to the Dreambooth Extension. 
I'm also curious to see if the training result would be better than just training images alone like the current dreambooth model.

**Reply by Jonseed** [reply]
@YukihimeX the extension already has the option to train with captions in filenames or txt files.

**Reply by YukihimeX** [reply]
> @Jonseed  the extension already has the option to train with captions in filenames or txt files.

Are you by chance referring to the [Textual Inversion](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Textual-Inversion) training method, perhaps? As for this [SD Dreambooth Extension](https://github.com/d8ahazard/sd_dreambooth_extension), I don't remember seeing it in the options (maybe I missed it or need an update).

**Reply by Jonseed** [reply]
@YukihimeX no, I mean this extension. Update it and the option is directly below the instance prompt field.

**Reply by YukihimeX** [reply]
> @Jonseed  no, I mean this extension. Update it and the option is directly below the instance prompt field.

Oh awesome... thanks for the kind info, appreciate it.

**Reply by wktra** [reply]
> n. Update it and the option is directly below the instance prompt field. @Jonseed 

how do we get it to read captions from text files? is it as simple as adding a text file of captions for each image with the same name as the corresponding image? (example: car(1).png, car(1).txt, car(2).png, car(2).txt, etc, etc)

**Reply by Jonseed** [reply]
@wktra the easiest way is to use the preprocess images function on the train tab. That will automatically produce caption files using BLIP or deepbooru, which you can then edit if you want. Or you can do it manually, and just create txt files that are named the same as the images.

**Reply by wktra** [reply]
> @wktra the easiest way is to use the preprocess images function on the train tab. That will automatically produce caption files using BLIP or deepbooru, which you can then edit if you want. Or you can do it manually, and just create txt files that are named the same as the images.

@Jonseed  Thank you. I've done that before when training with hypernetworks. 

But will dreambooth **automatically** read those text files next to the training images?  Is there some setting I have to enable?

**Reply by wktra** [reply]
@Jonseed also, does it matter whether we use the danbooru style of captioning within the text file? Or the clip style? 

`cat_ears,1girl, solo, cat_tail`

or 

`cat ears, girl, solo, cat tail`

**Reply by Jonseed** [reply]
@wktra I think you have to specify [filewords] in either the instance prompt or in your json config file for it to read the text files. I don't think it matters whether you use the danbooru style or clip style, but it probably depends on the base model you are using.

**Reply by wktra** [reply]
@Jonseed thank you!

@d8ahazard you're the authority on this and your opinion is MUCH appreciated and respected. 🙏🙏 When I'm supplying descriptive text files next to each training image, is it sufficient to use [filewords] in the instance prompt?

Do I type something like `sks, [filewords]` in that space?

**Reply by cerega66** [reply]
> I don't know if there's a definitive answer to this, but this is my thought process:
> 
> The classification images are sort of a baseline or control for what the new subject images should look like in the most generic sense possible. Specifically, what the existing model was creating before training started, and what it's creating now.
> 
> I've read a few guides that suggest just taking random datasets from the web and using those as classifiers, but from my most basic and unscientific tests, I haven't gotten great results this way.
> 
> In contrast, just generating the basic class images using the existing model seems to provide better editability and coherence.
> 
> Again, none of this is science, just me tinkering with one dataset, but from what I can tell, just let the prompt do whatever it's going to do.
> 
> Alternatively, I'd be curious to know what would happen if we generated a wider variety of class images with custom prompts embedded in the filename/txt, and using those prompts with the class image. I'll probably implement that at some point in time in the future....

I'll expand on your answer a bit.
When trained using a prompt class, they have a direct impact on how your model trains. I may have said that they create a framework from the promt class that your concept will not go beyond, but no. In fact, everything is easier. This can be described with the following formula:
class prompt=instance prompt=instance image + (class image * prior_loss_weight).
That is, you can put any images and they will affect in their own way how to learn your concept.

Further, only a theory that is quite difficult to confirm or disprove, but I came to it after spending a lot of experiments.
Using images generated by the model itself determines the possibilities and scope, but not the model itself, but just the model with its capabilities and limits your concept. But why is it better to use model images and not your own? Because every little detail affects how your concept learns, and the model generates quite a lot of small randomness in the images themselves. That is, the quality of the generated images directly affects training.

Now about the quantity. Since we understand how images affect, we can calculate their effective number. When learning by step according to the standard, 1 training image is compared with 1 class image. and in theory, with 100 training images and 100 steps per image, we need 100 class images. But this is provided that the training goes as follows:
1 step 1 train image 1 class image
2 step 2 train image 1 class image
....
100 step 100 train  image 1 class image
101 step 1 train image 2 class image
etc.

Or:
1 step 1 train image 1 class image
2 step 1 train image 2 class image
....
100 step 1 train image 100 class image
101 step 2 train image 1 class image
etc.

But if the training looks like:
1 step 1 train image 1 class image
2 step 2 train image 2 class image
....
100 step 100 train image 100 class image
101 step 1 train image 1 class image
etc.

then here 100 is not enough and in fact we need 100*100=10000.

That is, we get that the maximum number of class images at which they have efficiency = the number of steps.

**Reply by wktra** [reply]
> > I don't know if there's a definitive answer to this, but this is my thought process:
> > 
> > The classification images are sort of a baseline or control for what the new subject images should look like in the most generic sense possible. Specifically, what the existing model was creating before training started, and what it's creating now.
> > 
> > I've read a few guides that suggest just taking random datasets from the web and using those as classifiers, but from my most basic and unscientific tests, I haven't gotten great results this way.
> > 
> > In contrast, just generating the basic class images using the existing model seems to provide better editability and coherence.
> > 
> > Again, none of this is science, just me tinkering with one dataset, but from what I can tell, just let the prompt do whatever it's going to do.
> > 
> > Alternatively, I'd be curious to know what would happen if we generated a wider variety of class images with custom prompts embedded in the filename/txt, and using those prompts with the class image. I'll probably implement that at some point in time in the future....
> 
> I'll expand on your answer a bit.
> When trained using a prompt class, they have a direct impact on how your model trains. I may have said that they create a framework from the promt class that your concept will not go beyond, but no. In fact, everything is easier. This can be described with the following formula:
> class prompt=instance prompt=instance image + (class image * prior_loss_weight).
> That is, you can put any images and they will affect in their own way how to learn your concept.
> 
> Further, only a theory that is quite difficult to confirm or disprove, but I came to it after spending a lot of experiments.
> Using images generated by the model itself determines the possibilities and scope, but not the model itself, but just the model with its capabilities and limits your concept. But why is it better to use model images and not your own? Because every little detail affects how your concept learns, and the model generates quite a lot of small randomness in the images themselves. That is, the quality of the generated images directly affects training.
> 
> Now about the quantity. Since we understand how images affect, we can calculate their effective number. When learning by step according to the standard, 1 training image is compared with 1 class image. and in theory, with 100 training images and 100 steps per image, we need 100 class images. But this is provided that the training goes as follows:
> 1 step 1 train image 1 class image
> 2 step 2 train image 1 class image
> ....
> 100 step 100 train  image 1 class image
> 101 step 1 train image 2 class image
> etc.
> 
> Or:
> 1 step 1 train image 1 class image
> 2 step 1 train image 2 class image
> ....
> 100 step 1 train image 100 class image
> 101 step 2 train image 1 class image
> etc.
> 
> But if the training looks like:
> 1 step 1 train image 1 class image
> 2 step 2 train image 2 class image
> ....
> 100 step 100 train image 100 class image
> 101 step 1 train image 1 class image
> etc.
> 
> then here 100 is not enough and in fact we need 100*100=10000.
> 
> That is, we get that the maximum number of class images at which they have efficiency = the number of steps.

So if I follow this logic, then it's good to have the same number of class images as training images? 😵
